src/workerd/api/sockets.h

```diff
@@ -18,16 +18,30 @@ struct SocketAddress {
 
 struct SocketOptions {
   jsg::Unimplemented tls; // TODO(later): TCP socket options need to be implemented.
-  JSG_STRUCT(tls);
+  bool allowHalfOpen;
+  JSG_STRUCT(tls, allowHalfOpen);
 };
 
 class Socket: public jsg::Object {
 public:
-  Socket(jsg::Lock& js, jsg::Ref<ReadableStream> readable, jsg::Ref<WritableStream> writable,
-      kj::Own<jsg::PromiseResolverPair<void>> close)
-      : readable(kj::mv(readable)), writable(kj::mv(writable)),
+  Socket(jsg::Lock& js, jsg::Ref<ReadableStream> readableParam, jsg::Ref<WritableStream> writable,
+      kj::Own<jsg::PromiseResolverPair<void>> close, kj::Promise<void> connDisconnPromise,
+      jsg::Optional<SocketOptions> options)
+      : readable(kj::mv(readableParam)), writable(kj::mv(writable)),
         closeFulfiller(kj::mv(close)),
-        closedPromise(kj::mv(closeFulfiller->promise)) { };
+        closedPromise(kj::mv(closeFulfiller->promise)),
+        writeDisconnectedPromise(nullptr),
+        options(kj::mv(options)) {
+    // Listen for abrupt disconnects and resolve the `closed` promise when it occurs.
+    writeDisconnectedPromise = connDisconnPromise.then([this]() {
+      closeFulfiller->resolver.resolve();
+    }).eagerlyEvaluate(nullptr);
+
+    // Listen for EOF on the ReadableStream.
+    KJ_ASSERT_NONNULL(readable->eofResolverPair)->promise.then(js, [this](jsg::Lock& js) {
+      return closeWriteSide(js);
+    }).markAsHandled();
```

Comment:
This probably needs a JSG_VISITABLE_LAMBDA and a ref to this to be safe...

```suggestion
    KJ_ASSERT_NONNULL(readable->eofResolverPair)->promise.then(js,
        JSG_VISITABLE_LAMBDA((ref=JSG_THIS), (ref), (jsg::Lock& js) {
      return ref->closeWriteSide(js);
    })).markAsHandled();
```

The only problem there, of course, is that `JSG_THIS` does not work within the constructor of a `jsg::Object` :-/ ...

Reply:
Hm, it seems to be working. Would you expect it to fail to compile or is there a nasty bug hiding here that I did not manage to reproduce?

Reply:
Nasty bug it is. Reproduced it when running workerd (though oddly my tests didn't repro it)

Reply:
Fixed!